text/0000-multiarch-builders-and-package.md

```diff
@@ -0,0 +1,944 @@
+
+This seems to be the case for [Paketo Buildpacks](https://github.com/paketo-buildpacks/java/blob/main/package.toml) 
+or [Heroku](https://github.com/heroku/buildpacks-jvm/blob/main/meta-buildpacks/java/package.toml). A simplified version of their folder structures is:
+
+```bash
+├── bin
+│ ├── build
+│ └── detect
+├── buildpack.toml
+└── package.toml
+```
+
+In these cases, the expected output will be similar to:
+
+```bash
+pack buildpack package <buildpack> --config ./package.toml --publish 
+Warning: A new '--platform' flag is available to set the target platform for the buildpack package, using 'linux' as default
+Successfully published package <buildpack> and saved to registry
+
+# Or
+pack buildpack package  <buildpack> --config ./package.toml --format file 
+Warning: A new '--platform' flag is available to set the target platform for the buildpack package, using 'linux' as default
+Successfully created package <buildpack> and saved to file
+```
+
+We will keep `linux` as the default platform, and we don't expect end-users to change anything else on their side.
+
+In case, Buildpack authors tries to use the new folder structure like:
+
+```bash
+├── buildpack.toml
+├── linux-amd64
+│ └── bin
+│     ├── build
+│     └── detect
+├── linux-arm64
+│ └── bin
+│     ├── build
+│     └── detect
+└── package.toml
+```
+
+The current `pack` version creates an OCI artifact copying all the files on it. 
+
+![](https://hackmd.io/_uploads/BkGkZ_mla.png)
+
+But I think this buildpack can't be executed because `./bin/build` or `./bin/detect` can't be found. We could add some validations and throw an error,  
+but I do not see any value on doing that. 
+
+> **Important**
+> New folder structure is not useful if Buildpack Authors don't remove stacks and migrate to use targets
+> 
+
+Trying to use the new flags
+
+```bash
+pack buildpack package <buildpack> --config ./package.toml --publish --multi-arch
+Error: 'targets' or 'platforms' must be defined when creating a multi-architecture buildpack
+
+# Or
+pack buildpack package  <buildpack> --config ./package.toml --format file --multi-arch
+Error: 'targets' or 'platforms' must be defined when creating a multi-architecture buildpack
+```
+
+In these cases, we don't have enough information to create a multi-arch buildpack, and pack must fail its execution. 
+
+A valid way to re-write the command could be:
+
+```bash
+pack buildpack package <buildpack> --config ./package.toml --publish --platform linux/arm64 --platform linux/amd64 --multi-arch
+A multi-arch buildpack package will be created for platforms: 'linux/amd64', 'linux/arm64'
+Successfully published package <buildpack> and saved to registry
+
+# Or
+pack buildpack package  <buildpack> --config ./package.toml --format file --platform linux/arm64 --platform linux/amd64  --multi-arch
+A multi-arch buildpack package will be created for platforms: 'linux/amd64', 'linux/arm64'
+Successfully created package <buildpack> and saved to file
+```
```

Comment:
So this command could be used if one hasn't updated to use targets yet? i.e. still using stacks?

What about buildpacks that are migrating from stacks to no-stacks and have both present? This use case concerns me the most, cause I know we (Paketo) will be in this state for a while.

Reply:
I'd be interested to hear more in general about this state of migrating from stacks to no-stacks. What commands would use the old stack metadata? what commands would use the new target info? I'm assuming we'd keep the two synchronized, but just curious more about how this would all work.

Or am I overthinking this, and the net result (i.e. the images generated) will be the same regardless of stacks or no stacks?

Reply:
Good point, regarding the stack removal [plan](https://docs.google.com/document/d/1bExpr31U5R5yQ6fncpl5YcdosWVwYcXgkt12vE-lpvU/edit#heading=h.jpmej4uevvh8) we are planning to support both scenarios (stacks and run, build images), based on your feedback the idea will be to clarify how a world with:
- stacks
- targets
- multi-arch support 
will work